appctype/appconnector.go

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package appcfg contains an experimental configuration structure for
+// "tailscale.com/app-connector" capmap extensions.
+package appctype
+
+import (
+	"net/netip"
+
+	"tailscale.com/tailcfg"
+)
+
+// AppConnectorConfig is the configuration structure for an application
+// connection proxy service.
+type AppConnectorConfig struct {
```

Comment:
The WIP consuming code tracks and handles easy proxy 'unit' separately. As such, trying to scan through an array to work out what actually changed so it can start/stop listeners etc is more trouble than its worth.

Make this a map instead, where the key is any opaque value you want. You could also make this a map where the key is the 'listening' configuration, up to you.

Reply:
I'll defer to @maisem here who didn't seem to like the map

Reply:
what is a `proxy unit`? What would be the map key be?

Reply:
It'll probably correspond to a  `ServiceDomain`. Perhaps a nic_id or obfuscated equivalent.

It can be `string`, as an opaque value.

Reply:
Can you provide an example config? If you see sniProxy section in "TestGolden" this format helps trim what we send down on the wire.
The previous examples with a serviceID didn't handle a service that required multiple types of handlers, like it didn't allow you to do SNI based forwarding + TCP forwarding.

Reply:
If the aim is to minimize bytes, we can make DNAT and the forwarding enumeration a subset of each config unit:

```json
{
  "app": {
    "tailscale.com/app-connector": {
      "nic_afdas7654as87": {
        "sni_domains": {"10.64.1.1": "yeet.com"},
        "DNAT": {"10.64.1.1": ...},
        ".comment": "Addresses between DNAT and sni_domains may overlap in the same proxy unit."
      }
    }
  }
}
```

Reply:
>        "sni_domains": {"10.64.1.1": "yeet.com"},

You would also need the respective v6 addresses

In this example, there is also no way to specify a single config where sniProxy is enabled for all the addresses.

Reply:
A less intuitive approach that would save more bytes would be to split out listening addresses from DNAT and sni_domains, and just reference the listening addresses by index:

```json

      "nic_afdas7654as87": {
        "listening_addrs": ["10.64.1.1", "2002::1"],
        "sni_domains": {"yeet.com": [0, 1]},
        "DNAT": {},
      }
```

Reply:
Wouldnt that then loose the association between different types of handlers for the same service?

```json
"10.64.1.1": {
  "accept_sni_domains": ["yeet.com"],
  "tcp_ports": [22, 80]
}
```

So then we are back to loosing the 'service' information

Reply:
I meant you could do that on the client side, i.e. the wire format remains what James put in here.
And then you would iterate over `dnat[].addrs[]` and create an in-memory map to do the comparisons.

```
{
  "dnat": [{
    "addrs": ["100.64.0.1", "fd7a:115c:a1e0::1"],
    "to": ["example.org"],
    "ip": ["*"]
  }],
  "sniProxy": [{
    "addrs": ["::"],
    "ip": ["tcp:443"],
    "allowedDomains": ["*"]
  }],
  "advertiseRoutes": true
}
```

Apart from easing client side diff, is there any other reason we should pass down the `service` name down to the client?

Reply:
My issue with this approach is we've decomposed whats actually happening down into something that has lost semantic meaning - that usually ends up boxing us in a corner later

Sure we can probably work around the client diff issue with some terrible iteration that separates into new, changed, and deleted sets, but why not just keep this information from the get go?

DNAT proxying and SNI proxying are always subconfigs of some higher-level 'unit', why should we loose that grouping?

Reply:
If we really want to have SNIProxy and DNAT be the top-level can we at least add a field which groups them?

```json
{
  "dnat": [{
    "addrs": ["100.64.0.1", "fd7a:115c:a1e0::1"],
    "to": ["example.org"],
    "ip": ["*"],
    "group": "nic_fdggfddf"
  }],
  "sniProxy": [{
    "addrs": ["::"],
    "ip": ["tcp:443"],
    "allowedDomains": ["*"],
    "group": "nic_fdggfddf"
  }],
  "advertiseRoutes": true
}
```

Reply:
Thats probably a wider issue with the config: listeners should be associated with a service. Lets track the 'proxy all' SNI config separately to service configuration:

```json
{
  "dnat": [{
    "addrs": ["100.64.0.1", "fd7a:115c:a1e0::1"],
    "to": ["example.org"],
    "ip": ["*"],
    "group": "nic_1"
  },{
    "addrs": ["100.64.0.2", "fd7a:115c:a1e0::2"],
    "to": ["foo.com"],
    "ip": ["*"],
    "group": "nic_2"
  }],
  "sniProxy": [{
    "addrs": ["100.64.0.2", "fd7a:115c:a1e0::2"],
    "ip": ["tcp:443"],
    "allowedDomains": [ "foo.com"],
    "group": "nic_fdggfddf"
  }],
  "allowAllSNI": true
}
```

Reply:
Speaking of listeners, are you expecting the DNAT stuff to be listeners? that implies that you would need to run 65k*number of protocols*number of services.

Reply:
Listeners arent explicit netstack 'sockets' in tsnet - they are just mappings: https://github.com/tailscale/tailscale/blob/v1.50.1/tsnet/tsnet.go#L977

I think the listener array and associated infrastructure will likely need some refactoring, but I think we can be more efficient than having an entry for each address/port/protocol combination.

Reply:
> Listeners arent explicit netstack 'sockets' in tsnet - they are just mappings: https://github.com/tailscale/tailscale/blob/v1.50.1/tsnet/tsnet.go#L977

:nod:

>I think the listener array and associated infrastructure will likely need some refactoring, but I think we can be more efficient than having an entry for each address/port/protocol combination.


I think you might want to plumb through some other mechanism that allows you to intercept/alter packets directly instead of trying to listen on certain ports. IOW you probably don't even want to terminate the connection for the non-sni proxy cases.

Reply:
:nod:

Reply:
This is a long thread on something that we dont have any cost to change while we are developing. Lets just get this in?